### PR TITLE
convert code to TF 0.12

### PR DESCRIPTION
**Pull request by tibers**
Hey folks.

I did some quick munging to get this to TF 0.12 syntax. Specifically I want to call out moving 00-vars.tf to variables.tf and `required_provider` is broken and should be avoided. I also changed the JSON notation to be more TF readable for someone with their brain in TF.

### CODE REVIEW

**Comment by paemason** [comment]
Thanks for the help.

You remove `required_provider` but we hit issues with the Packet provider 2.8 so our fix was to lock it into 2.7.5. What is your suggestion to use the version of Packet provider we want if not use `required_provider`?

**Comment by tibers** [comment]
We're up to 2.10.0 for the packet provider, could you please run it in your test area and ensure that the issue is fixed to your satisfaction?

File: variables.tf [hunk 1]
```diff
@@ -144,7 +141,7 @@ variable "gcs_bucket_name" {
 }
 
 variable "s3_url" {
-  default = "https://s3.example.com"
+  default = "https=//s3.example.com"
```
Comment (displague) [inline variables.tf]: This seems like a typo. @tibers?

Comment (tibers) [reply variables.tf]: Good catch! Thank you!

**Comment by paemason** [comment]
We continue to see issues related to converting from l3 to hybrid mode when using any packet terraform provider beyond 2.7.5.

Until we can reliably use the later provider with c2, c3, m2, and n2 machines we'll need to stay with 2.7.5.

I've posted my issues with the packet terraform provider into the slack channel.

File: variables.tf [hunk 2]
```diff
@@ -27,7 +24,7 @@ variable "project_id" {
 }
 
 /*
-Valid vsphere_service_types are:
+Valid vsphere_service_types are=
```
Comment (displague) [inline variables.tf:27]: stray colon/equal change

**Review by displague — APPROVED** [review]
Other than the nitpick on the stray colon/equal conversion, this looks good to me.

For added style clean-up, I think `terraform fmt` would be safe to include in this PR.

**Comment by displague** [comment]
As long as we are adapting to TF0.12, each of the lines that appear in `git grep -n '"${'  | grep .tf:` create TF12 warnings.

The `"${` and trailing `}"` are safe to remove in most of these lines.

Some lines would take extra care are.

### String Concatenation

```
01-create-project.tf:5:  ssh_key_name = "${local.project_name_sanitized}-${local.timestamp_sanitized}-key"
anthos/static-ip.tf:116:      addresses: ["${var.ipv4_address}/${var.ipv4_netmask_prefix_length}"]
```

### Nested evaluation

```
01-create-project.tf:3:  timestamp_sanitized = "${replace("${local.timestamp}", "/[-| |T|Z|:]/", "")}"
01-create-project.tf:4:  project_name_sanitized = "${replace("${var.project_name}", "/[ ]/","_")}"
```

### Simple changes

```
$ git grep -n '"${'  | grep .tf:
01-create-project.tf:2:  timestamp           = "${timestamp()}"
anthos/static-ip.tf:66:  user           = "${var.vsphere_user}"
anthos/static-ip.tf:67:  password       = "${var.vsphere_password}"
anthos/static-ip.tf:68:  vsphere_server = "${var.vsphere_server}"
anthos/static-ip.tf:77:  name          = "${var.datastore}"
anthos/static-ip.tf:78:  datacenter_id = "${data.vsphere_datacenter.dc.id}"
anthos/static-ip.tf:82:  name = "${var.datacenter}"
anthos/static-ip.tf:86:  name          = "${var.cluster}"
anthos/static-ip.tf:87:  datacenter_id = "${data.vsphere_datacenter.dc.id}"
anthos/static-ip.tf:91:  name          = "${var.resource_pool}"
anthos/static-ip.tf:92:  datacenter_id = "${data.vsphere_datacenter.dc.id}"
anthos/static-ip.tf:96:  name          = "${var.network}"
anthos/static-ip.tf:97:  datacenter_id = "${data.vsphere_datacenter.dc.id}"
anthos/static-ip.tf:101:  name          = "${var.vm_template}"
anthos/static-ip.tf:102:  datacenter_id = "${data.vsphere_datacenter.dc.id}"
anthos/static-ip.tf:140:    static_ip_config = "${base64encode(data.template_file.static_ip_config.rendered)}"
anthos/static-ip.tf:142:    reg_enable = "${var.registry_enable}"
anthos/static-ip.tf:143:    reg_username = "${var.registry_username}"
anthos/static-ip.tf:144:    reg_password = "${var.registry_password}"
anthos/static-ip.tf:145:    reg_dns_hostname = "${var.registry_dns_hostname}"
anthos/static-ip.tf:154:    # reg_htpasswd = "${file(var.registry_htpasswd)}"
anthos/static-ip.tf:155:    # reg_cert = "${file(var.registry_cert)}"
anthos/static-ip.tf:156:    # reg_private_key = "${file(var.registry_private_key)}"
anthos/static-ip.tf:157:    # reg_ca = "${file(var.registry_ca)}"
anthos/static-ip.tf:168:  name             = "${var.vm_name}"
anthos/static-ip.tf:169:  resource_pool_id = "${data.vsphere_resource_pool.pool.id}"
anthos/static-ip.tf:170:  datastore_id     = "${data.vsphere_datastore.datastore.id}"
anthos/static-ip.tf:171:  num_cpus         = "${var.num_cpus}"
anthos/static-ip.tf:172:  memory           = "${var.memory}"
anthos/static-ip.tf:173:  guest_id         = "${data.vsphere_virtual_machine.template_from_ovf.guest_id}"
anthos/static-ip.tf:175:  scsi_type = "${data.vsphere_virtual_machine.template_from_ovf.scsi_type}"
anthos/static-ip.tf:177:    network_id   = "${data.vsphere_network.network.id}"
anthos/static-ip.tf:178:    adapter_type = "${data.vsphere_virtual_machine.template_from_ovf.network_interface_types[0]}"
anthos/static-ip.tf:188:    size             = "${max(50, data.vsphere_virtual_machine.template_from_ovf.disks.0.size)}"
anthos/static-ip.tf:189:    eagerly_scrub    = "${data.vsphere_virtual_machine.template_from_ovf.disks.0.eagerly_scrub}"
anthos/static-ip.tf:190:    thin_provisioned = "${data.vsphere_virtual_machine.template_from_ovf.disks.0.thin_provisioned}"
anthos/static-ip.tf:199:      hostname    = "${var.vm_name}"
anthos/static-ip.tf:200:      public-keys = "${file(var.ssh_public_key_path)}"
anthos/static-ip.tf:201:      user-data   = "${base64encode(data.template_file.user_data.rendered)}"
anthos/static-ip.tf:206:    template_uuid = "${data.vsphere_virtual_machine.template_from_ovf.id}"
anthos/static-ip.tf:211:  value = "${vsphere_virtual_machine.vm.default_ip_address}"
```

**Comment by tibers** [comment]
@displague I ran `terraform fmt` in the PR to include your formatting requests - this also fixes the instances of `${}` variables not needing that format if they're not being interpolated.

**Comment by displague** [comment]
I successfully provisioned with this branch merged locally.

**Comment by tibers** [comment]
bumping for attention - any further interest in this?

**Comment by displague** [comment]
@paemason What do you think of these changes? We have a few PRs open like this that are mostly stylistic.  I've approved the ones that I've successfully deployed with (in a local merging branch).

I'd like to see this repository eventually become [published in the Terraform Registry](https://www.terraform.io/docs/registry/modules/publish.html), [perhaps even certified](https://www.terraform.io/docs/registry/modules/verified.html).  This comes with [some best practice recommendations](https://www.terraform.io/docs/modules/#standard-module-structure).

> main.tf, variables.tf, outputs.tf. These are the recommended filenames for a minimal module, even if they're empty. main.tf should be the primary entrypoint. For a simple module, this may be where all the resources are created. For a complex module, resource creation may be split into multiple files but any nested module calls should be in the main file. variables.tf and outputs.tf should contain the declarations for variables and outputs, respectively.

@tibers It looks like this needs a rebase to pick up the `00-vars.tf` -> `variables.tf` changes.  Also, per these recommendations, `99-output.tf` should be renamed to `outputs.tf` and a stub `main.tf` should be added (empty is fine).

**Comment by paemason** [comment]
In general we need to gain maturity around testing and review of PRs against this repository. For my part, I'm mostly contributing to the functionality of the Anthos portions of the code. I would expect that the Packet team have oversight and decision making over the packet terraform portions and direction of the code.

**Comment by displague** [comment]
Thanks for that clarification, @paemason. I didn't want to step on your toes.  I really love how this project has evolved and gained momentum.  I'm happy to assist in that effort!

**Comment by displague** [comment]
Merged via [2d9b5f8..f29eca6](https://github.com/packet-labs/google-anthos/compare/2d9b5f8..f29eca6) - Thanks @tibers!

**Pull request closed without merging**